clodius/tiles/bigbed.py

```diff
@@ -0,0 +1,407 @@
+                   'tile_size': 1024,
+                   'max_zoom': 7
+                  }
+    '''
+    TILE_SIZE = 1024
+
+    if chromsizes is None:
+        chromsizes = get_chromsizes(bbpath)
+        chromsizes_list = []
+
+        for chrom, size in chromsizes.iteritems():
+            chromsizes_list += [[chrom, int(size)]]
+    else:
+        chromsizes_list = chromsizes
+
+    min_tile_cover = np.ceil(
+        sum([int(c[1]) for c in chromsizes_list]) / TILE_SIZE
+    )
+    max_zoom = int(np.ceil(np.log2(min_tile_cover)))
+
+    tileset_info = {
+        'min_pos': [0],
+        'max_pos': [TILE_SIZE * 2 ** max_zoom],
+        'max_width': TILE_SIZE * 2 ** max_zoom,
+        'tile_size': TILE_SIZE,
+        'max_zoom': max_zoom,
+        'chromsizes': chromsizes_list,
+        'range_modes': range_modes
+    }
+    return tileset_info
+
+
+def fetch_data(a):
+    (
+        bbpath,
+        binsize,
+        chromsizes,
+        range_mode,
+        min_elements,
+        max_elements,
+        cid,
+        start,
+        end
+    ) = a
+    n_bins = int(np.ceil((end - start) / binsize))
+
+    try:
+        chrom = chromsizes.index[cid]
+        clen = chromsizes.values[cid]
+
+        kwargs = {"bins": n_bins, "missing": np.nan}
+        
+        fetch_factory = ft.partial(bbi.fetch_intervals, bbpath, chrom, start, end)
+
+        if range_mode == 'significant':
+            intervals, intervals2 = fetch_factory(), fetch_factory()
+
+        else:
+            intervals, intervals2 = fetch_factory(), fetch_factory()
+
+    except IndexError:
+        # beyond the range of the available chromosomes
+        # probably means we've requested a range of absolute
+        # coordinates that stretch beyond the end of the genome
+        intervals, intervals2 = None, None
+        
+    except KeyError:
+        # probably requested a chromosome that doesn't exist (e.g. chrM)
+        intervals, intervals2 = None, None
+        
+    offset = 0
+    offsetIdx = 0
+    chrOffsets = {}
+    for chrSize in chromsizes:
+        chrOffsets[chromsizes.index[offsetIdx]] = offset
+        offset += chrSize
+        offsetIdx += 1
+        
+    final_intervals = []
+    intervals_length = 0
+    scores = []
+    
+    if not intervals:
+        return final_intervals
+    
+    for interval in intervals:
+        scores.append(int(interval[4]))
+        intervals_length += 1
+    
+    # generate beddb-like elements for parsing by the higlass plugin
+    if intervals_length >= min_elements and intervals_length <= max_elements:
+        for interval in intervals2:
+            final_intervals.append({'chrOffset': chrOffsets[chrom], 'importance': int(interval[4]), 'fields': interval})
+            
+    elif intervals_length > max_elements:
+        thresholded_intervals = []
+        desired_perc = max_elements / intervals_length
+        thresholded_score = int(np.quantile(scores, 1-desired_perc))
+        for interval in intervals2:
```

Comment:
Might be clearer to use a list comprehension here:

```
thresholded_intervals = [interval for interval in intervals2 if int(interval[4]) >= thresholded_score]
```

Reply:
I used a list comprehension, slightly modified to return `beddb`-like elements.

Reply:
I do not know of a way to do try/except handling within a list comprehension, so in case the bigBed data lacks score data, or score data are not integer-like, I reverted back to a `for` loop in order to handle that case.